src/viam/sdk/common/proto_value.hpp

```diff
@@ -0,0 +1,324 @@
+        // operations on other provided by other_vtable.
+        void swap(const vtable& this_vtable,
+                  storage& other,
+                  const vtable& other_vtable) noexcept(impl::all_proto_moves_noexcept{});
+
+        // Destroy this storage, using the destructor from vtable.
+        void destruct(const vtable& vtable) noexcept;
+
+        template <typename T = void>
+        T* get() {
+            return static_cast<T*>(static_cast<void*>(&buf_));
+        }
+
+        template <typename T = void>
+        T const* get() const {
+            return static_cast<T const*>(static_cast<void const*>(&buf_));
+        }
+
+        BufType buf_;
+    };
+
+    ProtoValue(const google::protobuf::Value* value);
+
+    vtable vtable_;
+    storage self_;
+};
+
+/// @brief Alias declaration for map of string to type-erased ProtoValue representing
+/// google::protobuf::Struct.
+/// This stores structured data as a map where the keys can be thought of as member names.
+using ProtoStruct = std::unordered_map<std::string, ProtoValue>;
+
+void to_proto(std::nullptr_t, google::protobuf::Value* v);
+void to_proto(bool b, google::protobuf::Value* v);
+void to_proto(int i, google::protobuf::Value* v);
+void to_proto(double d, google::protobuf::Value* v);
+void to_proto(std::string s, google::protobuf::Value* v);
+void to_proto(const std::vector<ProtoValue>& vec, google::protobuf::Value* v);
+void to_proto(const ProtoStruct& m, google::protobuf::Value* v);
+void to_proto(const ProtoValue& t, google::protobuf::Value* v);
+
+ProtoStruct struct_to_map(google::protobuf::Struct const* s);
+void map_to_struct(const ProtoStruct& m, google::protobuf::Struct* s);
+
+/// @brief Convert a type to a google::protobuf::Value.
+/// @note This method is trivially templated to insulate google::protobuf::Value from our
+/// API/ABI. It is meant to be called with no template parameters in a translation unit which
+/// includes <google/protobuf/struct.pb.h>
+template <typename T, typename Value = google::protobuf::Value>
+Value to_proto(T&& t) {
+    Value v;
+    to_proto(std::forward<T>(t), &v);
+
+    return v;
+}
+
+/// @brief Convert a google::protobuf::Struct to a ProtoStruct.
+/// @note This method is trivially templated to insulate google::protobuf::Struct from our
+/// API/ABI. It is meant to be called with no template parameters in a translation unit which
+/// includes <google/protobuf/struct.pb.h>
+template <typename Struct = google::protobuf::Struct>
+ProtoStruct struct_to_map(const Struct& s) {
+    return struct_to_map(&s);
+}
+
+/// @brief Convert a ProtoStruct to a google::protobuf::Struct.
+/// @note This method is trivially templated to insulate google::protobuf::Struct from our
+/// API/ABI. It is meant to be called with no template parameters in a translation unit which
+/// includes <google/protobuf/struct.pb.h>
+template <typename Struct = google::protobuf::Struct>
+Struct map_to_struct(const ProtoStruct& m) {
```

Comment:
These `map_to_struct` and `struct_to_map` names are the same as those for the existing `ProtoType`. Is there going to be a conflict/ambiguity if they are both in scope? Or are they intended to be drop-in replacements as part of a strategy of rolling over to this implementation? I was never much of a fan of the names using `map` anyway.

Reply:
Answering this and the question above about roll-out strategy:

For a couple reasons I think this may need to be a one-fell-swoop approach:
- Old `ProtoType` and `AttributeMap` code is littered with `std::make_shared` and lots of pointer dereferencing, null checks, etc. I think this would make a wrapper approach more trouble than it's worth.
- A lot of use of `AttributeMap` originates from the root of an inheritance hierarchy with a lot of leaves, e.g., `do_command` in all the component classes, which then is coupled with the corresponding server implementations.

That said, adding `map_to_struct` and `struct_to_map` was motivated by an audit of the existing code, and in those cases the replacement might be a drop-in, modulo a pointer deref in some cases